File: ios/FitnessGeolocation/DebugMonitor.swift
import Foundation
import UIKit
import AudioToolbox
import UserNotifications
import os.log

// MARK: - Motion State Machine

enum MotionState: String {
  case moving, stationary
}

/// Binary motion state machine with stopTimeout hysteresis.
/// Mirrors transistorsoft's detection: moving ↔ stationary with 5-min timeout.
class MotionStateMachine {
  var state: MotionState = .stationary
  var stopTimeoutMinutes: TimeInterval = 5
  var currentActivity: String = "unknown"
  var currentConfidence: Double = 0
  var stateSince: Date = Date()
  var stopTimerRemaining: TimeInterval = 0

  private var stopTimer: Timer?
  private var stationarySince: Date?
  weak var delegate: MotionStateMachineDelegate?

  func feedActivity(_ activity: String, confidence: Double, timestamp: Date) {
    currentActivity = activity
    currentConfidence = confidence

    switch activity {
    case "walking", "running", "cycling", "driving":
      transitionTo(.moving, timestamp: timestamp)
    case "stationary", "unknown":
      startStopTimeout(timestamp: timestamp)
    default:
      break
    }
  }

  func feedSpeed(_ speed: Double, timestamp: Date) {
    if speed > 0.5 {
      // Speed indicates movement
      transitionTo(.moving, timestamp: timestamp)
    }
  }

  private func startStopTimeout(timestamp: Date) {
    if state == .stationary { return } // Already stationary
    if stopTimer != nil { return } // Timer already running

    stopTimer = Timer.scheduledTimer(withTimeInterval: stopTimeoutMinutes * 60, repeats: false) { [weak self] _ in
      guard let self = self else { return }
      self.stopTimer = nil
      self.stopTimerRemaining = 0
      self.transitionTo(.stationary, timestamp: Date())
      self.delegate?.motionStateMachine(self, didFire: "stop_timeout_start", message: "Stop timeout elapsed — now stationary")
      self.delegate?.motionStateMachine(self, didPlaySound: "stop_timeout_start")
    }
    stopTimerRemaining = stopTimeoutMinutes * 60

    delegate?.motionStateMachine(self, didFire: "stop_timeout_start", message: "Device still — stop timeout \(Int(stopTimeoutMinutes))min started")
    delegate?.motionStateMachine(self, didPlaySound: "stop_timeout_start")
  }

  func cancelStopTimeout() {
    guard stopTimer != nil else { return }
    stopTimer?.invalidate()
    stopTimer = nil
    stopTimerRemaining = 0
    delegate?.motionStateMachine(self, didFire: "stop_timeout_cancel", message: "Device moved — stop timeout cancelled")
    delegate?.motionStateMachine(self, didPlaySound: "stop_timeout_cancel")
  }

  private func transitionTo(_ newState: MotionState, timestamp: Date) {
    guard state != newState else {
      // Cancel stop timeout if we're moving and timer was running
      if newState == .moving { cancelStopTimeout() }
      return
    }

    let oldState = state
    state = newState
    stateSince = timestamp

    if newState == .moving {
      cancelStopTimeout()
      stationarySince = nil
    } else {
      stationarySince = timestamp
    }

    delegate?.motionStateMachine(self, didChangeState: newState, from: oldState, activity: currentActivity)
    delegate?.motionStateMachine(self, didPlaySound: newState == .moving ? "motionchange_true" : "motionchange_false")

    let msg = newState == .moving
      ? "Started moving — \(currentActivity)"
      : "Stopped — now stationary"
    delegate?.motionStateMachine(self, didFire: "motionchange", message: msg)
  }

  func reset() {
    stopTimer?.invalidate()
    stopTimer = nil
    stopTimerRemaining = 0
    state = .stationary
    stateSince = Date()
    stationarySince = nil
    currentActivity = "unknown"
  }

  func toDictionary() -> [String: Any] {
    [
      "state": state.rawValue,
      "activity": currentActivity,
      "confidence": currentConfidence,
      "sinceTimestamp": stateSince.timeIntervalSince1970 * 1000,
      "stopTimeoutRemaining": stopTimerRemaining,
    ]
  }
}

protocol MotionStateMachineDelegate: AnyObject {
  func motionStateMachine(_ machine: MotionStateMachine, didChangeState state: MotionState, from oldState: MotionState, activity: String)
  func motionStateMachine(_ machine: MotionStateMachine, didFire event: String, message: String)
  func motionStateMachine(_ machine: MotionStateMachine, didPlaySound sound: String)
}

// MARK: - Heartbeat Engine

class HeartbeatEngine {
  var intervalSeconds: TimeInterval = 60
  private var timer: Timer?
  weak var delegate: HeartbeatEngineDelegate?

  func start() {
    stop()
    timer = Timer.scheduledTimer(withTimeInterval: intervalSeconds, repeats: true) { [weak self] _ in
      guard let self = self else { return }
      self.delegate?.heartbeatEngine(self, didHeartbeat: [
        "event": "heartbeat",
        "message": "Heartbeat",
        "timestamp": Date().timeIntervalSince1970 * 1000,
      ])
      self.delegate?.heartbeatEngine(self, didPlaySound: "heartbeat")
    }
  }

  func stop() {
    timer?.invalidate()
    timer = nil
  }
}

protocol HeartbeatEngineDelegate: AnyObject {
  func heartbeatEngine(_ engine: HeartbeatEngine, didHeartbeat event: [String: Any])
  func heartbeatEngine(_ engine: HeartbeatEngine, didPlaySound sound: String)
}

// MARK: - Debug Notification Manager

class DebugNotificationManager {
  private var currentActivityText: String = "Stationary"
  private let oslog = OSLog(subsystem: "com.fitnessgeolocation", category: "debug-notif")

  func updateActivityText(_ text: String) {
    currentActivityText = text
  }

  func getActivityText() -> String { currentActivityText }

  func postLocalNotification(title: String, body: String) {
    let content = UNMutableNotificationContent()
    content.title = title
    content.body = body
    content.sound = nil // We play our own sounds

    let request = UNNotificationRequest(
      identifier: "com.fitnessgeolocation.debug.\(UUID().uuidString)",
      content: content,
      trigger: nil // Immediate
    )

    UNUserNotificationCenter.current().add(request) { error in
      if let error = error {
        os_log(.error, log: self.oslog, "notification_error: %@", error.localizedDescription)
      }
    }
  }

  func updateAndroidNotification(text: String) {
    // This is a no-op on iOS — Android handles it natively
  }
}

// MARK: - Sound Manager

class DebugSoundManager {
  private let oslog = OSLog(subsystem: "com.fitnessgeolocation", category: "debug-sound")
  var soundEnabled = true

  /// System sound IDs mapped to lifecycle events
  private let soundMap: [String: UInt32] = [
    "motionchange_true": 1016,    // Received Message (short chime)
    "motionchange_false": 1075,   // Mail Sent (descending tone)
    "location_recorded": 1104,    // SMS Received (short click)
    "location_error": 1006,       // Voice mail (error-like)
    "heartbeat": 1105,            // SMS Sent (subtle tick)
    "geofence_enter": 1023,       // Received Message (different pitch)
    "geofence_exit": 1075,        // Mail Sent
    "stop_timeout_start": 1004,   // Begin Recording (alert-like)
    "stop_timeout_cancel": 1070,  // Key Press (cancel-like)
    "stop_detection_delay": 1110, // SMS Received 2
  ]

  func play(_ sound: String) {
    guard soundEnabled else { return }
    if let soundId = soundMap[sound] {
      AudioServicesPlaySystemSound(soundId)
      os_log(.debug, log: oslog, "sound: %@ (id: %d)", sound, soundId)
    }
  }
}

// MARK: - DebugMonitor (Orchestrator)

class DebugMonitor: NSObject {
  let stateMachine = MotionStateMachine()
  let heartbeat = HeartbeatEngine()
  let notifications = DebugNotificationManager()
  let sounds = DebugSoundManager()

  weak var delegate: DebugMonitorDelegate?

  private var _enabled = false
  private let oslog = OSLog(subsystem: "com.fitnessgeolocation", category: "debug-monitor")

  // Activity → notification text mapping
  private var notificationTexts: [String: String] = [
    "stationary": "Stationary",
    "walking": "Walking",
    "running": "Running",
    "cycling": "Cycling",
    "driving": "Driving",
    "unknown": "Stationary",
    "moving": "Moving",
  ]

  var enabled: Bool {
    get { _enabled }
    set {
      let changed = _enabled != newValue
      _enabled = newValue
      if changed {
        delegate?.debugMonitor(self, didChangeEnabled: newValue)
        emitLifecycle("enabledChange", newValue ? "Debug monitoring enabled" : "Debug monitoring disabled")
      }
    }
  }

  override init() {
    super.init()
    stateMachine.delegate = self
    heartbeat.delegate = self
  }

  func configure(config: [String: Any]) {
    if let stopTimeout = config["stopTimeoutMinutes"] as? NSNumber {
      stateMachine.stopTimeoutMinutes = stopTimeout.doubleValue
    }
    if let heartbeatInterval = config["heartbeatIntervalSeconds"] as? NSNumber {
      heartbeat.intervalSeconds = heartbeatInterval.doubleValue
    }
    if let sound = config["sound"] as? Bool {
      sounds.soundEnabled = sound
    }

    // Notification text templates
    if let text = config["notificationTextStationary"] as? String { notificationTexts["stationary"] = text }
    if let text = config["notificationTextWalking"] as? String { notificationTexts["walking"] = text }
    if let text = config["notificationTextRunning"] as? String { notificationTexts["running"] = text }
    if let text = config["notificationTextCycling"] as? String { notificationTexts["cycling"] = text }
    if let text = config["notificationTextDriving"] as? String { notificationTexts["driving"] = text }
    if let text = config["notificationTextMoving"] as? String { notificationTexts["moving"] = text }
    if let title = config["notificationTitle"] as? String {
      UserDefaults.standard.set(title, forKey: "notification_title")
    }

    let wasEnabled = enabled
    enabled = config["enabled"] as? Bool ?? wasEnabled

    if enabled {
      heartbeat.start()
      emitLifecycle("configured", "Debug monitor configured")
    } else {
      heartbeat.stop()
    }

    os_log(.debug, log: oslog, "configured enabled=%d stopTimeout=%.1fmin heartbeat=%.1fs",
           enabled, stateMachine.stopTimeoutMinutes, heartbeat.intervalSeconds)
  }

  /// Feed motion activity into the state machine
  func feedActivity(_ activity: String, confidence: Double) {
    guard enabled else { return }
    stateMachine.feedActivity(activity, confidence: confidence, timestamp: Date())
    updateNotificationText(for: activity)
  }

  /// Feed speed data into the state machine
  func feedSpeed(_ speed: Double) {
    guard enabled else { return }
    stateMachine.feedSpeed(speed, timestamp: Date())
  }

  func getMotionState() -> [String: Any] {
    stateMachine.toDictionary()
  }

  func reset() {
    stateMachine.reset()
    heartbeat.stop()
    enabled = false
  }

  private func updateNotificationText(for activity: String) {
    let text: String
    switch activity {
    case "walking": text = notificationTexts["walking"] ?? "Walking"
    case "running": text = notificationTexts["running"] ?? "Running"
    case "cycling": text = notificationTexts["cycling"] ?? "Cycling"
    case "driving": text = notificationTexts["driving"] ?? "Driving"
    case "stationary": text = notificationTexts["stationary"] ?? "Stationary"
    default:
      text = stateMachine.state == .moving
        ? (notificationTexts["moving"] ?? "Moving")
        : (notificationTexts["stationary"] ?? "Stationary")
    }
    notifications.updateActivityText(text)
  }

  private func emitLifecycle(_ event: String, _ message: String, data: [String: Any] = [:]) {
    var payload: [String: Any] = data
    payload["event"] = event
    payload["message"] = message
    payload["timestamp"] = Date().timeIntervalSince1970 * 1000
    delegate?.debugMonitor(self, didEmitLifecycleEvent: payload)
  }
}

// MARK: - Delegates

extension DebugMonitor: MotionStateMachineDelegate {
  func motionStateMachine(_ machine: MotionStateMachine, didChangeState state: MotionState, from oldState: MotionState, activity: String) {
    let payload = machine.toDictionary()
    delegate?.debugMonitor(self, didEmitMotionState: payload)
    emitLifecycle("motionStateChange", "State: \(state.rawValue) — \(activity)", data: payload)
    updateNotificationText(for: activity)
  }

  func motionStateMachine(_ machine: MotionStateMachine, didFire event: String, message: String) {
    emitLifecycle(event, message)
    if event == "stop_timeout_start" || event == "stop_timeout_cancel" {
      notifications.postLocalNotification(title: "FitnessGeolocation", body: message)
    }
  }

  func motionStateMachine(_ machine: MotionStateMachine, didPlaySound sound: String) {
    sounds.play(sound)
  }
}

extension DebugMonitor: HeartbeatEngineDelegate {
  func heartbeatEngine(_ engine: HeartbeatEngine, didHeartbeat event: [String: Any]) {
    delegate?.debugMonitor(self, didEmitHeartbeat: event)
  }

  func heartbeatEngine(_ engine: HeartbeatEngine, didPlaySound sound: String) {
    sounds.play(sound)
  }
}

File: android/src/main/java/com/fitnessgeolocation/DebugMonitor.kt
package com.fitnessgeolocation

import android.content.Context
import android.media.RingtoneManager
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.PowerManager
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import java.util.concurrent.ConcurrentLinkedQueue

// ─── Motion State Machine ───────────────────────────────────────────────────

enum class MotionState { moving, stationary }

class MotionStateMachine {
  var state: MotionState = MotionState.stationary
  var stopTimeoutMinutes: Long = 5
  var currentActivity: String = "unknown"
  var currentConfidence: Double = 0.0
  var stateSince: Long = System.currentTimeMillis()
  var stopTimerRemaining: Long = 0

  private var stopTimer: Runnable? = null
  private val handler = Handler(Looper.getMainLooper())
  var delegate: MotionStateMachineDelegate? = null

  fun feedActivity(activity: String, confidence: Double) {
    currentActivity = activity
    currentConfidence = confidence

    when (activity) {
      "walking", "running", "cycling", "driving" -> transitionTo(MotionState.moving)
      "stationary", "unknown" -> startStopTimeout()
    }
  }

  fun feedSpeed(speed: Float) {
    if (speed > 0.5f && state == MotionState.stationary) {
      transitionTo(MotionState.moving)
    }
  }

  private fun startStopTimeout() {
    if (state == MotionState.stationary) return
    if (stopTimer != null) return

    stopTimerRemaining = stopTimeoutMinutes * 60
    val runnable = Runnable {
      stopTimer = null
      stopTimerRemaining = 0
      transitionTo(MotionState.stationary)
      delegate?.onEvent("stop_timeout_start", "Stop timeout elapsed — now stationary")
      delegate?.onSound("stop_timeout_start")
    }
    stopTimer = runnable
    handler.postDelayed(runnable, stopTimeoutMinutes * 60 * 1000)

    delegate?.onEvent("stop_timeout_start", "Device still — stop timeout ${stopTimeoutMinutes}min started")
    delegate?.onSound("stop_timeout_start")
  }

  fun cancelStopTimeout() {
    stopTimer?.let { handler.removeCallbacks(it) }
    stopTimer = null
    stopTimerRemaining = 0
    delegate?.onEvent("stop_timeout_cancel", "Device moved — stop timeout cancelled")
    delegate?.onSound("stop_timeout_cancel")
  }

  private fun transitionTo(newState: MotionState) {
    if (state == newState) {
      if (newState == MotionState.moving) cancelStopTimeout()
      return
    }

    val oldState = state
    state = newState
    stateSince = System.currentTimeMillis()

    if (newState == MotionState.moving) {
      cancelStopTimeout()
    }

    delegate?.onStateChange(newState, oldState, currentActivity)
    delegate?.onSound(if (newState == MotionState.moving) "motionchange_true" else "motionchange_false")

    val msg = if (newState == MotionState.moving) "Started moving — $currentActivity" else "Stopped — now stationary"
    delegate?.onEvent("motionchange", msg)
  }

  fun reset() {
    stopTimer?.let { handler.removeCallbacks(it) }
    stopTimer = null
    stopTimerRemaining = 0
    state = MotionState.stationary
    stateSince = System.currentTimeMillis()
    currentActivity = "unknown"
  }

  fun toMap(): Map<String, Any?> = mapOf(
    "state" to state.name,
    "activity" to currentActivity,
    "confidence" to currentConfidence,
    "sinceTimestamp" to stateSince.toDouble(),
    "stopTimeoutRemaining" to stopTimerRemaining.toDouble(),
  )
}

interface MotionStateMachineDelegate {
  fun onStateChange(newState: MotionState, oldState: MotionState, activity: String)
  fun onEvent(event: String, message: String)
  fun onSound(sound: String)
}

// ─── Heartbeat Engine ───────────────────────────────────────────────────────

class HeartbeatEngine {
  var intervalSeconds: Long = 60
  var delegate: HeartbeatEngineDelegate? = null

  private val handler = Handler(Looper.getMainLooper())
  private var runnable: Runnable? = null

  fun start() {
    stop()
    val r = object : Runnable {
      override fun run() {
        delegate?.onHeartbeat(mapOf(
          "event" to "heartbeat",
          "message" to "Heartbeat",
          "timestamp" to System.currentTimeMillis().toDouble(),
        ))
        delegate?.onSound("heartbeat")
        handler.postDelayed(this, intervalSeconds * 1000)
      }
    }
    runnable = r
    handler.postDelayed(r, intervalSeconds * 1000)
  }

  fun stop() {
    runnable?.let { handler.removeCallbacks(it) }
    runnable = null
  }
}

interface HeartbeatEngineDelegate {
  fun onHeartbeat(event: Map<String, Any?>)
  fun onSound(sound: String)
}

// ─── Sound Manager ──────────────────────────────────────────────────────────

class DebugSoundManager(private val context: Context) {
  var soundEnabled = true

  fun play(sound: String) {
    if (!soundEnabled) return
    try {
      val uri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
      val ringtone = RingtoneManager.getRingtone(context, uri)
      ringtone.play()
      Log.d("FitnessGeoSound", "sound: $sound")
    } catch (e: Exception) {
      Log.w("FitnessGeoSound", "sound_failed: ${e.message}")
    }
  }
}

// ─── DebugMonitor (Orchestrator) ────────────────────────────────────────────

interface DebugMonitorDelegate {
  fun onEnabledChange(enabled: Boolean)
  fun onMotionState(state: Map<String, Any?>)
  fun onHeartbeatEvent(event: Map<String, Any?>)
  fun onLifecycleEvent(event: Map<String, Any?>)
}

class DebugMonitor(private val context: Context) {
  val stateMachine = MotionStateMachine()
  val heartbeat = HeartbeatEngine()
  val sounds = DebugSoundManager(context)

  var delegate: DebugMonitorDelegate? = null

  private var _enabled = false
  private val tag = "FitnessGeoDebug"

  // Activity → notification text mapping
  private var notificationTexts = mutableMapOf(
    "stationary" to "Stationary",
    "walking" to "Walking",
    "running" to "Running",
    "cycling" to "Cycling",
    "driving" to "Driving",
    "unknown" to "Stationary",
    "moving" to "Moving",
  )
  private var notificationTitle = "Tracking activity"

  var enabled: Boolean
    get() = _enabled
    set(value) {
      val changed = _enabled != value
      _enabled = value
      if (changed) {
        delegate?.onEnabledChange(value)
        emitLifecycle("enabledChange", if (value) "Debug monitoring enabled" else "Debug monitoring disabled")
      }
    }

  init {
    stateMachine.delegate = object : MotionStateMachineDelegate {
      override fun onStateChange(newState: MotionState, oldState: MotionState, activity: String) {
        val payload = stateMachine.toMap()
        delegate?.onMotionState(payload)
        emitLifecycle("motionStateChange", "State: ${newState.name} — $activity")
        updateNotificationText(activity)
      }

      override fun onEvent(event: String, message: String) {
        emitLifecycle(event, message)
      }

      override fun onSound(sound: String) {
        sounds.play(sound)
      }
    }

    heartbeat.delegate = object : HeartbeatEngineDelegate {
      override fun onHeartbeat(event: Map<String, Any?>) {
        delegate?.onHeartbeatEvent(event)
      }

      override fun onSound(sound: String) {
        sounds.play(sound)
      }
    }
  }

  fun configure(config: Map<String, Any?>) {
    config["stopTimeoutMinutes"]?.let { stateMachine.stopTimeoutMinutes = (it as Number).toLong() }
    config["heartbeatIntervalSeconds"]?.let { heartbeat.intervalSeconds = (it as Number).toLong() }
    config["sound"]?.let { sounds.soundEnabled = it as Boolean }

    config["notificationTextStationary"]?.let { notificationTexts["stationary"] = it.toString() }
    config["notificationTextWalking"]?.let { notificationTexts["walking"] = it.toString() }
    config["notificationTextRunning"]?.let { notificationTexts["running"] = it.toString() }
    config["notificationTextCycling"]?.let { notificationTexts["cycling"] = it.toString() }
    config["notificationTextDriving"]?.let { notificationTexts["driving"] = it.toString() }
    config["notificationTextMoving"]?.let { notificationTexts["moving"] = it.toString() }
    config["notificationTitle"]?.let { notificationTitle = it.toString() }

    enabled = config["enabled"] as? Boolean ?: enabled

    if (enabled) {
      heartbeat.start()
      emitLifecycle("configured", "Debug monitor configured")
    } else {
      heartbeat.stop()
    }

    Log.d(tag, "configured enabled=$enabled stopTimeout=${stateMachine.stopTimeoutMinutes}min heartbeat=${heartbeat.intervalSeconds}s")
  }

  fun feedActivity(activity: String, confidence: Double) {
    if (!enabled) return
    stateMachine.feedActivity(activity, confidence)
    updateNotificationText(activity)
  }

  fun feedSpeed(speed: Float) {
    if (!enabled) return
    stateMachine.feedSpeed(speed)
  }

  fun getMotionState(): Map<String, Any?> = stateMachine.toMap()

  fun reset() {
    stateMachine.reset()
    heartbeat.stop()
    enabled = false
  }

  private fun updateNotificationText(activity: String) {
    val text = when (activity) {
      "walking" -> notificationTexts["walking"] ?: "Walking"
      "running" -> notificationTexts["running"] ?: "Running"
      "cycling" -> notificationTexts["cycling"] ?: "Cycling"
      "driving" -> notificationTexts["driving"] ?: "Driving"
      "stationary" -> notificationTexts["stationary"] ?: "Stationary"
      else -> if (stateMachine.state == MotionState.moving)
        notificationTexts["moving"] ?: "Moving"
      else
        notificationTexts["stationary"] ?: "Stationary"
    }

    // Store for FitnessLocationService to read
    context.getSharedPreferences("fitness_geolocation", Context.MODE_PRIVATE)
      .edit()
      .putString("notification_text", text)
      .putString("notification_title", notificationTitle)
      .apply()
  }

  private fun emitLifecycle(event: String, message: String, data: Map<String, Any?> = emptyMap()) {
    val payload = mutableMapOf<String, Any?>()
    payload.putAll(data)
    payload["event"] = event
    payload["message"] = message
    payload["timestamp"] = System.currentTimeMillis().toDouble()
    delegate?.onLifecycleEvent(payload)

    Log.d(tag, "[$event] $message")
  }
}
